File: apps/cli/src/commands/upsert.ts
import { Command, flags } from "@oclif/command"
import fs from "fs"
import type { definition, platform } from "@uesio/ui"
import { uploadFiles } from "../upsert/uploadfiles"
import { printWorkspace } from "../print/workspace"
import { authorize } from "../auth/login"
import { getWorkspace, getApp } from "../config/config"
import { post } from "../request/request"
import { BodyInit } from "node-fetch"

export default class Upsert extends Command {
	static description = "upsert data"

	static flags = {
		spec: flags.string({ char: "s" }),
		file: flags.string({ char: "f" }),
		collection: flags.string({ char: "c" }),
	}

	static args = []

	async run(): Promise<void> {
		const { flags } = this.parse(Upsert)

		if (!flags.file) {
			console.log("No file specified", flags)
			return
		}

		const spec: definition.Spec = flags.spec
			? JSON.parse(await fs.promises.readFile(flags.spec, "utf8"))
			: {
					jobtype: "IMPORT",
			  }

		if (flags.collection) {
			spec.collection = flags.collection
		}

		let payload: BodyInit = ""

		if (spec.jobtype === "IMPORT") {
			if (!spec.collection) {
				console.log("No collection specified", spec)
				return
			}
			if (!spec.filetype) {
				spec.filetype = "CSV"
			}

			console.log("Upserting...", flags.file, spec)

			//const filedata = await fs.readFile(flags.file, "utf8")
			payload = fs.createReadStream(flags.file, {
				encoding: "utf8",
			})
		}

		if (spec.jobtype === "UPLOADFILES") {
			payload = await uploadFiles(flags.file, spec)
		}

		if (!payload) {
			throw new Error("Invalid jobtype")
		}

		const app = await getApp()
		const workspace = await getWorkspace()
		if (!workspace) {
			printWorkspace(app, workspace)
			return
		}

		const user = await authorize()

		const jobResponse = await post(
			`workspace/${app}/${workspace}/bulk/job`,
			JSON.stringify(spec),
			user.cookie
		)
		const jobResponseObj =
			(await jobResponse.json()) as platform.JobResponse
		// Create a batch
		const jobId = jobResponseObj.id
		const batchResponse = await post(
			`workspace/${app}/${workspace}/bulk/job/${jobId}/batch`,
			payload,
			user.cookie
		)
		if (batchResponse.status === 200) {
			const batchResponseObj =
				(await batchResponse.json()) as platform.JobResponse
			console.log(batchResponseObj.id)
		} else {
			const errorMessage = await batchResponse.text()
			console.log(errorMessage)
		}
	}
}

File: apps/clio/pkg/command/upsert.go
package command

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"

	"github.com/thecloudmasters/clio/pkg/auth"
	"github.com/thecloudmasters/clio/pkg/call"
	"github.com/thecloudmasters/clio/pkg/config"
	"github.com/thecloudmasters/clio/pkg/config/ws"
	"github.com/thecloudmasters/clio/pkg/zip"
	"github.com/thecloudmasters/uesio/pkg/bulk"
	"github.com/thecloudmasters/uesio/pkg/meta"
)

type UpsertOptions struct {
	SpecFile   string
	DataFile   string
	Collection string
}

func createJob(app, workspace, sessid string, spec *meta.JobSpec) (*bulk.JobResponse, error) {
	url := fmt.Sprintf("workspace/%s/%s/bulk/job", app, workspace)
	jobResponse := &bulk.JobResponse{}

	err := call.PostJSON(url, sessid, spec, jobResponse)
	if err != nil {
		return nil, err
	}

	fmt.Println("Created Job: " + jobResponse.ID)
	return jobResponse, nil
}

func getImportPayload(jobType, dataFile string) (io.Reader, error) {
	if jobType == "IMPORT" {
		return os.Open(dataFile)
	}
	if jobType == "UPLOADFILES" {
		return zip.ZipDir(dataFile), nil
	}
	return nil, errors.New("Invalid Job Type: " + jobType)
}

func runBatch(app, workspace, sessid, dataFile, jobID string, spec *meta.JobSpec) (*bulk.BatchResponse, error) {
	payload, err := getImportPayload(spec.JobType, dataFile)
	if err != nil {
		return nil, err
	}

	fmt.Println("Upserting...", dataFile)

	url := fmt.Sprintf("workspace/%s/%s/bulk/job/%s/batch", app, workspace, jobID)

	resp, err := call.Request("POST", url, payload, sessid)
	if err != nil {
		return nil, err
	}

	batchResponse := &bulk.BatchResponse{}

	err = json.NewDecoder(resp.Body).Decode(batchResponse)
	if err != nil {
		return nil, err
	}

	fmt.Println("Upsert Success: " + resp.Status)

	defer resp.Body.Close()
	return batchResponse, nil

}

func getSpec(options *UpsertOptions) (*meta.JobSpec, error) {
	spec := &meta.JobSpec{}

	if options.SpecFile != "" {
		f, err := os.Open(options.SpecFile)
		if err != nil {
			return nil, err
		}
		err = json.NewDecoder(f).Decode(spec)
		if err != nil {
			return nil, err
		}
	} else {
		spec.JobType = "IMPORT"
	}

	if options.Collection != "" {
		spec.Collection = options.Collection
	}

	if spec.Collection == "" {
		return nil, errors.New("No collection specified")
	}

	if spec.JobType == "IMPORT" {

		if spec.FileType == "" {
			spec.FileType = "CSV"
		}
	}
	return spec, nil
}

func Upsert(options *UpsertOptions) error {

	if options == nil {
		options = &UpsertOptions{}
	}

	if options.DataFile == "" {
		return errors.New("No Data File Specified")
	}

	fmt.Println("Running Upsert Command")

	_, err := auth.Login()
	if err != nil {
		return err
	}

	app, err := config.GetApp()
	if err != nil {
		return err
	}

	workspace, err := ws.GetWorkspace()
	if err != nil {
		return err
	}

	sessid, err := config.GetSessionID()
	if err != nil {
		return err
	}

	spec, err := getSpec(options)
	if err != nil {
		return err
	}

	jobResponse, err := createJob(app, workspace, sessid, spec)
	if err != nil {
		return err
	}

	batchResponse, err := runBatch(app, workspace, sessid, options.DataFile, jobResponse.ID, spec)
	if err != nil {
		return err
	}

	fmt.Println("Success Running Batch: " + batchResponse.ID)

	return nil
}
